<a href="https://colab.research.google.com/github/amatchison/CS4410/blob/main/Homework7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [73]:
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D

In [66]:
#create neural network
cnn = Sequential()

In [67]:
#add layers
cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Flatten())

cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

In [68]:
#compile model
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [69]:
#load fasion_mnist
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

#reshape data
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

#normalize
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

#one-hot encoding
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [70]:
#train model
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 109s 128ms/step - loss: 0.4646 - accuracy: 0.8338 - val_loss: 0.3271 - val_accuracy: 0.8835
Epoch 2/5
844/844 [==============================] - 106s 126ms/step - loss: 0.3020 - accuracy: 0.8908 - val_loss: 0.2934 - val_accuracy: 0.8915
Epoch 3/5
844/844 [==============================] - 109s 129ms/step - loss: 0.2551 - accuracy: 0.9064 - val_loss: 0.2597 - val_accuracy: 0.9042
Epoch 4/5
844/844 [==============================] - 109s 129ms/step - loss: 0.2222 - accuracy: 0.9179 - val_loss: 0.2576 - val_accuracy: 0.9027
Epoch 5/5
844/844 [==============================] - 109s 129ms/step - loss: 0.1947 - accuracy: 0.9269 - val_loss: 0.2509 - val_accuracy: 0.9070


In [71]:
#evaluate model
loss, accuracy = cnn.evaluate(X_test, y_test)

313/313 [==============================] - 7s 21ms/step - loss: 0.2693 - accuracy: 0.9017


In [72]:
#load mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

#reshape data
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

#normalize
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

#one-hot encoding
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

11490434/11490434 [==============================] - 0s 0us/step


In [63]:
#train model
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 107s 127ms/step - loss: 0.1730 - accuracy: 0.9358 - val_loss: 0.2548 - val_accuracy: 0.9088
Epoch 2/5
844/844 [==============================] - 107s 127ms/step - loss: 0.1513 - accuracy: 0.9445 - val_loss: 0.2264 - val_accuracy: 0.9182
Epoch 3/5
844/844 [==============================] - 109s 130ms/step - loss: 0.1293 - accuracy: 0.9509 - val_loss: 0.2391 - val_accuracy: 0.9170
Epoch 4/5
844/844 [==============================] - 106s 125ms/step - loss: 0.1098 - accuracy: 0.9596 - val_loss: 0.2716 - val_accuracy: 0.9143
Epoch 5/5
844/844 [==============================] - 107s 126ms/step - loss: 0.0966 - accuracy: 0.9632 - val_loss: 0.2575 - val_accuracy: 0.9172


In [65]:
#evaluate model
loss, accuracy = cnn.evaluate(X_test, y_test)

313/313 [==============================] - 5s 18ms/step - loss: 0.2923 - accuracy: 0.9129




---


# How well does the model perform on Fashion-MNIST compared to MNIST?
The model performs similarly on both data sets. Fashion-MNIST had a loss of 0.27 and an accuracy of 0.90. MNIST had 0.29 loss and 0.91 accuracy.


# How do the training times compare?
Both had similar training times of around 110s per epoch.